cpp/docs/DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,977 @@
+```c++
+template<cudf::type_id t> struct always_int{ using type = int32_t; }
+
+// This will always invoke `operator()<int32_t>`
+cudf::type_dispatcher<always_int>(data_type, f);
+```
+
+## Specializing Type-Dispatched Code Paths
+
+It is often necessary to customize the dispatched `operator()` for different types. This can be 
+done in several ways.
+
+The first method is to use explicit, full template specialization. This is useful for specializing 
+behavior for single types. The following example function object prints `"int32_t"` or `"double"` 
+when invoked with either of those types, or `"unhandled type"` otherwise.
+
+```c++
+struct type_printer {
+template <typename ColumnType>
+void operator()() { std::cout << "unhandled type\n"; }
+};
+
+// Due to a bug in g++, explicit member function specializations need to be
+// defined outside of the class definition
+template <>
+void type_printer::operator()<int32_t>() { std::cout << "int32_t\n"; }
+
+template <>
+void type_printer::operator()<double>() { std::cout << "double\n"; }
+```
+
+The second method is to use [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae) with 
+`std::enable_if_t`. This is useful to partially specialize for a set of types with a common trait. 
+The following example functor prints `integral` or `floating point` for integral or floating point
+types, respectively.
+
+```c++
+struct integral_or_floating_point {
+template <typename ColumnType,
+          std::enable_if_t<not std::is_integral<ColumnType>::value and
+                           not std::is_floating_point<ColumnType>::value>* = nullptr> 
+void operator()() { std::cout << "neither integral nor floating point\n"; }
+
+template <typename ColumnType,
+          std::enable_if_t<std::is_integral<ColumnType>::value>* = nullptr>
+void operator()() { std::cout << "integral\n"; }
+
+template < typename ColumnType,
+           std::enable_if_t<std::is_floating_point<ColumnType>::value>* = nullptr> 
+void operator()() { std::cout << "floating point\n"; }
+};
+```
+
+For more info on SFINAE with `std::enable_if`, [see this post](https://eli.thegreenplace.net/2014/sfinae-and-enable_if).
+
+There are a number of traits defined in `include/cudf/utilities/traits.hpp` that are useful for 
+partial specialization of dispatched function objects. For example `is_numeric<T>()` can be used to 
+specialize for any numeric type.
+
+# Nested Data Types
+
+libcudf supports a number of nested data types, including strings, lists, and structs. A string is 
+simply a character string, but a column of strings may have a different-length string in each row.
+A list is a list of elements of any type, so a column of lists of integers has rows with a list of 
+integers, possibly of a different length, in each row. In a colum of structs, each row is a 
+structure comprising one or more fields. These fields are stored in structure-of-arrays format, so 
+that the column of structs has a nested column for each field of the structure. 
```

Comment:
Might we consider converting this paragraph to a list? (No pun intended.) 
Also, fixed typo in `colum`.
```suggestion
libcudf supports a number of nested data types, including strings, lists, and structs: 
1. `String`: This is simply a character string, but a column of strings may have a 
different-length string in each row.
2. `List`: This is a list of elements of any type, so a column of lists of integers has 
rows with a list of integers, possibly of a different length, in each row. 
3. `Struct`: In a column of structs, each row is a structure comprising one or more 
fields. These fields are stored in structure-of-arrays format, so that the column of 
structs has a nested column for each field of the structure. 
```